### ISSUE TITLE
Use precompiled headers

### ISSUE THREAD

**Issue by TheJJ** [SFTtech/openage#501]
To speed up things especially with templates, gcc supports header precompiling.

https://gcc.gnu.org/onlinedocs/gcc/Precompiled-Headers.html

Please integrate this into our buildsystem! Header files would have to be registered like `.cpp` files then.
##

**Comment by TheJJ** [comment]
`clang` and `gcc` currently throw warnings when a header with `#pragma once` is compiled. We have to wait for an upstream fix.

**Comment by zuntrax** [comment]
Someone is working on a fix for llvm: https://llvm.org/bugs/show_bug.cgi?id=16686

**Comment by zuntrax** [comment]
`clang` stopped emitting this kind of warning. `gcc` still has it. I've seen some shell hackery on SO that circumvents this kind of warning, but I want to make sure those headers behave the same. http://stackoverflow.com/a/21740973

**Comment by zuntrax** [comment]
OK. So my gcc segfaults when I try this trick. Not very nice.
But there is another way:
```
echo '#include "test.h"' > test.h_wrapper
g++ -x c++-header --std=c++14 -Wall -Wextra -pedantic -o test.h.gch test.h_wrapper
rm test.h_wrapper
```

If we hide this abomination somewhere deep in CMake where nobody will ever look, this might be fine.

**Comment by zuntrax** [comment]
We had a chat in matrix and discussed how we could use precompiled headers. There are some serious technical limitations, especially only being able to use one (1) such header per translation unit. The only reasonable way to handle this seems to be precompiling one central file which includes all commonly used headers and including that everywhere. Using `-include` flags leads to formally incorrect header files which suddenly don't build when PCH is off. Manually including would further increase compile times without PCH, which is bad as well.
Long story short, probably not worth the trouble.

**Comment by mic-e** [comment]
As a final note, C++17 modules (which were voted out of C++17 and will remain a TS until C++20) are basically "PCHs on steroids, among other things" and would pretty much remove every single header file from the project. Which would have been completely awesome. Modules are currently supported by LLVM and MSVC. Maybe it would be worth moving to them as soon as GCC supports them as well.
